pkg/tcpip/transport/tcp/endpoint.go

```diff
@@ -1138,11 +1148,8 @@ func (e *endpoint) Peek(vec [][]byte) (int64, tcpip.ControlMessages, *tcpip.Erro
 //
 // It must be called with rcvListMu held.
 func (e *endpoint) zeroReceiveWindow(scale uint8) bool {
```

Comment:
Shouldn't this also account for cases where scaling is disabled? Don't we want this to return true if the available space goes below that of MSS?

Reply:
window scaling: I removed it from test, it's irrelvant here.

Negative. We keep the semantics of zero-window exactly as they were. Instead we redefine the semantics of what should happen on recv() when window is small but non-zero.

In past: recv() would trigger ACK _only_ when window was in zero-window mode.
Now: recv() will do that, but also trigger ACK when window grew to >= 1MSS.

The difference is subtle, and probably wrong - although it's another discussion if we want to fix this. Consider recv() returning small value in three cases:
 - (A) recv()=small when window is zero
 - (B) recv()=small when window is one
 - (C) recv()=mss when window is one

Before this patch (A) would generate ACK, (B),(C) would not. With this patch (A) and (C) genrate ack, (B) does not.

Arguably, we could implement fully recv side silly window syndrome avoidance and prevent (A) from generating ACK if the avail is small.... not sure if this is part of this bug/stream of work.

Reply:
Won't the correct behaviour be for A and B to not generate a window update since there isn't enough space available? I believe only C) should generate a window update right to avoid silly-window. Eventually a future read should clear enough space and cause a zero window to be generated due to > MSS space being available right?

Reply:
Correct, but I'm not totally sure we want to address (A) in this PR. Handling (A) might require some more guarantees (like ensuring recvbufSize > MSS).
Also, current semantics are not wrong, they may be just considered wasteful (spurious acks). For (C) which this PR addresses the semantics as they stand now are plain wrong.

Reply:
Agreed. I am fine with this change as-is and we can address A separately.